container/memory_store.go

```diff
@@ -73,19 +84,60 @@ func (c *memoryStore) First(filter StoreFilter) *Container {
 // ApplyAll calls the reducer function with every container in the store.
 // This operation is asyncronous in the memory store.
 func (c *memoryStore) ApplyAll(apply StoreReducer) {
-	c.Lock()
-	defer c.Unlock()
+	c.globalLock.Lock()
+	defer c.globalLock.Unlock()
 
 	wg := new(sync.WaitGroup)
 	for _, cont := range c.s {
 		wg.Add(1)
+		c.namedLock.Lock(cont.ID)
 		go func(container *Container) {
+			defer func() {
+				c.namedLock.Unlock(container.ID)
+				wg.Done()
+			}()
 			apply(container)
-			wg.Done()
 		}(cont)
 	}
 
 	wg.Wait()
 }
 
+// ReduceAll filters a list of containers and calls the reducer function with each one of them.
+func (c *memoryStore) ReduceAll(filter StoreFilter, apply StoreReducer) error {
+	var containers []*Container
+	c.globalLock.Lock()
+	for _, cont := range c.s {
+		if filter(cont) {
+			containers = append(containers, cont)
+		}
+	}
+	c.globalLock.Unlock()
+
+	for _, cont := range containers {
+		c.namedLock.Lock(cont.ID)
+		if err := apply(cont); err != nil {
+			c.namedLock.Unlock(cont.ID)
+			return err
```

Comment:
Is there any chance of this causing a problem if we have an error halfway through the containers, and then have half with stuff applied and half without?